### DISCUSSION TITLE
parameter verify function

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#244] [Design & Development]
In working on pwmig I realized there was an essential need for a generic function to verify parametric data.   I was about to sit down and write one for pwmig, but realized it would make sense to put it in mspass since it has generic use.   I'm posting this in our discussions to get feedback on an API.   

What is needed?

1. Is a key-value pair defined in a pf (or yaml) file defined?
2. Are the values the reasonable when converted to the type expected?
3.  Support for nested content (subdocuments in mongo = dict with dict values) which pf defines as &Arr and &Tbl.   

I was thinking of this originally as a function that would be run on an mspasspy.ccore.utility.AntelopePf object.  There are, I know, several other ways to define the same data structures.  Is there a clean way to do that for other ways of defining parameters like a mongodb document?

I think the verify process should be driven by content stored either in a data file or in MongoDB.   Although this is probably better done as python class, I can write the c++ version of what I think the abstraction of the constraints we should apply are:
```
class ParameterConstraints
{
public:   // This is really a struct because I'm making all its attributes public
  MDtype type;
  bool is_valid();   // generic method returning true if a parameter has a valid value
  // constructors, operator=, etc needed here
};

class ParameterTester
{
public:
   /* constructors - would require some thought.  Needs to load a full set of parameters with constraints that define
  constraints for a particular algorithm*/ 
  ParameterTester(args);
  /*! Test an AntelopePf for validity against constraints defined in this class */
  ErrorLogger testpf(const AntelopePf& pf);
 private:
  /* map container keyed by strings of keys that link to ParameterConstraints*/ 
  map<string,ParameterConstraints> constraints;
};
```

That is very incomplete, but illustrates one generic way to handle this - through the ParameterConstraints::is_valid method.  What I don't see yet is how to make this work with subdocument type parameter structures of any level.   

I'm going to post this for now to hear reactions and ideas before moving forward on this.

**Comment by pavlis** [comment]
A related issue comes up as I work through what I need for pwmig;  validate nonstandard (mspass nonstandard that is) data files.   A nearly universal thing a user will face, particularly in adapting legacy code to mspass, is reading some arcane file format containing data that program needs to run.  In many cases in the research environment that happens because the user is using old data from a previous project to continue the same work or build on it to do something new - a very very common research need in any field of science for anyone who isn't a student.  

I think in this situation we can assume the user is responsible for producing their own custom reader.   What we do need to consider providing is a generic test for existence and validity of the data.   That said, I suggest this basic signature for a generic file verify function:
```
def check_file(filename,directory=None,reader=None):
  """
  Checks for existence and (optionally) validity of an external data file.  

  :param filename:  file name to be tested.
  :param dir:   (optional) directory where filename is found.  Default assumes the file is in the current directory.  
  :param reader:  function to apply to filename to verify data.   May be a class constructor using a file name or the reader your main code will use to read this file.
```
I note it probably would make sense to add variable args to pass to reader so the reader could receive more than just a filename.  I don't exactly remember how to specify that in python.

**Reply by haruming** [reply]
The function signature looks good to me. However, for the parameter **reader** in the function, the comment says it could be a class constructor using a file name, which means it could be a class object as well?
For a function receiving more than defined parameters, one could simply use `def func(arg1, *args, k1=v1, **kwargs).`
